File: suc/swift/049.swift
import SwiftUI

// MARK: - AppTokens
// 翻译自 Kotlin 的 AppTokens 对象，用于集中管理应用的颜色、字体、形状、间距和阴影。
// 这种结构使得样式修改更加方便和统一。

struct AppTokens {
    struct Colors {
        // Kotlin Color(0xFFFF7043) -> Swift Color(red: 255/255.0, green: 112/255.0, blue: 67/255.0)
        static let primary = Color(red: 0xFF / 255.0, green: 0x70 / 255.0, blue: 0x43 / 255.0)
        static let secondary = Color(red: 0xFF / 255.0, green: 0xB7 / 255.0, blue: 0x4D / 255.0)
        static let tertiary = Color(red: 0xFF / 255.0, green: 0xD1 / 255.0, blue: 0x80 / 255.0)
        static let background = Color(red: 0xFF / 255.0, green: 0xF8 / 255.0, blue: 0xF2 / 255.0)
        static let surface = Color(red: 0xFF / 255.0, green: 0xFF / 255.0, blue: 0xFF / 255.0)
        static let surfaceVariant = Color(red: 0xFF / 255.0, green: 0xE0 / 255.0, blue: 0xB2 / 255.0)
        static let outline = Color(red: 0xD7 / 255.0, green: 0xCC / 255.0, blue: 0xC8 / 255.0)
        static let success = Color(red: 0x43 / 255.0, green: 0xA0 / 255.0, blue: 0x47 / 255.0)
        static let warning = Color(red: 0xFF / 255.0, green: 0xB3 / 255.0, blue: 0x00 / 255.0)
        static let error = Color(red: 0xE5 / 255.0, green: 0x39 / 255.0, blue: 0x35 / 255.0)
        static let onPrimary = Color(red: 0xFF / 255.0, green: 0xFF / 255.0, blue: 0xFF / 255.0)
        static let onSecondary = Color(red: 0x3E / 255.0, green: 0x27 / 255.0, blue: 0x23 / 255.0)
        static let onTertiary = Color(red: 0x3E / 255.0, green: 0x27 / 255.0, blue: 0x23 / 255.0)
        static let onBackground = Color(red: 0x3E / 255.0, green: 0x27 / 255.0, blue: 0x23 / 255.0)
        static let onSurface = Color(red: 0x3E / 255.0, green: 0x27 / 255.0, blue: 0x23 / 255.0)
    }

    struct TypographyTokens {
        // Kotlin sp 单位直接映射为 SwiftUI 的 pt (CGFloat)
        static let display = Font.system(size: 28, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular) // Kotlin's FontWeight.Normal 对应 SwiftUI 的 .regular
        static let label = Font.system(size: 12, weight: .medium)
    }

    struct Shapes {
        // Kotlin Dp 单位直接映射为 SwiftUI 的 CGFloat
        static let small: CGFloat = 8
        static let medium: CGFloat = 14
        static let large: CGFloat = 20
    }

    struct Spacing {
        // Kotlin Dp 单位直接映射为 SwiftUI 的 CGFloat
        static let sm: CGFloat = 6
        static let md: CGFloat = 10
        static let lg: CGFloat = 16
        static let xl: CGFloat = 24
        static let xxl: CGFloat = 36
    }

    // 阴影规范，用于精确映射 Kotlin 的 ShadowSpec
    struct ShadowSpec {
        let elevation: CGFloat // Kotlin 中的 elevation，在 SwiftUI 中通常通过 radius 和 y 偏移来模拟
        let radius: CGFloat    // 阴影的模糊半径
        let dy: CGFloat        // 阴影的 Y 轴偏移
        let opacity: Double    // 阴影颜色的不透明度

        var shadowColor: Color {
            Color.black.opacity(opacity)
        }
    }

    struct ElevationMapping {
        static let level1 = ShadowSpec(elevation: 2, radius: 4, dy: 2, opacity: 0.1)
        static let level2 = ShadowSpec(elevation: 6, radius: 8, dy: 4, opacity: 0.15)
        static let level3 = ShadowSpec(elevation: 10, radius: 12, dy: 6, opacity: 0.18)
    }
    
    // LazyVGrid 的最小项宽度，对应 Kotlin 的 GridCells.Adaptive(160.dp)
    static let gridItemMinSize: CGFloat = 160
}

// MARK: - Data Model
// 翻译自 Kotlin 的 TripCard 数据类

struct TripCard: Identifiable {
    let id: Int
    let title: String
    let days: Int
    let price: String
}

// MARK: - TripCardView
// 辅助视图，用于渲染单个行程卡片，保持 RootScreen 的整洁。
// 对应 Kotlin 代码中的 Card Composable。

struct TripCardView: View {
    let trip: TripCard

    var body: some View {
        VStack(alignment: .leading, spacing: AppTokens.Spacing.sm) {
            // 图像占位符 Box，对应 Kotlin 中的 Box with background and shape
            RoundedRectangle(cornerRadius: AppTokens.Shapes.medium)
                .fill(AppTokens.Colors.surfaceVariant)
                .frame(maxWidth: .infinity) // 对应 Modifier.fillMaxWidth()
                .frame(height: 100)          // 对应 Modifier.height(100.dp)

            Text(trip.title)
                .font(AppTokens.TypographyTokens.title)
                .foregroundColor(AppTokens.Colors.onSurface)

            Text("\(trip.days) days")
                .font(AppTokens.TypographyTokens.body)
                .foregroundColor(AppTokens.Colors.onSurface.opacity(0.7)) // 对应 .copy(alpha = 0.7f)

            HStack { // 对应 Kotlin 的 Row
                Text(trip.price)
                    .font(AppTokens.TypographyTokens.title)
                    .foregroundColor(AppTokens.Colors.primary)

                Spacer() // 对应 Kotlin 的 Arrangement.SpaceBetween

                Button(action: {
                    // 按钮点击事件，此处仅打印日志
                    print("Details for \(trip.title)")
                }) {
                    Text("Details")
                        .font(AppTokens.TypographyTokens.label)
                        .foregroundColor(AppTokens.Colors.onPrimary)
                        // 按钮内部填充，为了原子级对应 Kotlin 按钮的视觉效果
                        .padding(.horizontal, AppTokens.Spacing.md)
                        .padding(.vertical, AppTokens.Spacing.sm / 2) // 调整垂直填充以达到 36.dp 的高度
                }
                .background(AppTokens.Colors.primary) // 对应 ButtonDefaults.buttonColors(containerColor = ...)
                .cornerRadius(AppTokens.Shapes.medium) // 对应 shape = AppTokens.Shapes.medium
                .frame(height: 36) // 对应 Modifier.height(36.dp)
            }
        }
        .padding(AppTokens.Spacing.md) // 卡片内容的内部填充，对应 Column 的 padding
        .background(AppTokens.Colors.surface) // 卡片背景色，对应 CardDefaults.cardColors
        .cornerRadius(AppTokens.Shapes.large) // 卡片圆角，对应 shape = AppTokens.Shapes.large
        .shadow(color: AppTokens.ElevationMapping.level2.shadowColor, // 对应 CardDefaults.cardElevation
                radius: AppTokens.ElevationMapping.level2.radius,
                x: 0, // Kotlin 的 elevation 通常只影响 Y 轴偏移，X 轴默认为 0
                y: AppTokens.ElevationMapping.level2.dy)
    }
}

// MARK: - RootScreen
// 翻译自 Kotlin 的 RootScreen Composable，是应用的主要 UI 布局。

struct RootScreen: View {
    // 数据列表，对应 Kotlin 中的 remember { listOf(...) }
    let trips: [TripCard] = [
        TripCard(id: 1, title: "Kyoto Cherry Trail", days: 4, price: "$460"),
        TripCard(id: 2, title: "Tokyo City Break", days: 3, price: "$390"),
        TripCard(id: 3, title: "Osaka Gourmet Tour", days: 5, price: "$520"),
        TripCard(id: 4, title: "Mount Fuji Escape", days: 2, price: "$280"),
        TripCard(id: 5, title: "Okinawa Beach Week", days: 6, price: "$740"),
        TripCard(id: 6, title: "Hokkaido Winter Lights", days: 5, price: "$680")
    ]

    var body: some View {
        // 对应 Kotlin 的 Scaffold，背景渐变覆盖整个屏幕，内容在 VStack 中布局。
        VStack(alignment: .leading, spacing: AppTokens.Spacing.md) { // 对应 Kotlin Column 的 verticalArrangement
            Text("Itinerary Planner")
                .font(AppTokens.TypographyTokens.display)
                .foregroundColor(AppTokens.Colors.primary)
                .padding(.horizontal, AppTokens.Spacing.lg) // 标题的水平填充，与整体内容对齐

            ScrollView { // 对应 Kotlin 的 LazyVerticalGrid 外层的可滚动区域
                LazyVGrid(
                    columns: [GridItem(.adaptive(minimum: AppTokens.gridItemMinSize), spacing: AppTokens.Spacing.lg)], // 对应 GridCells.Adaptive 和 horizontalArrangement
                    spacing: AppTokens.Spacing.lg // 对应 verticalArrangement
                ) {
                    ForEach(trips) { trip in // 对应 items(trips)
                        TripCardView(trip: trip)
                    }
                }
                .padding(.horizontal, AppTokens.Spacing.lg) // Grid 的水平填充，与整体内容对齐
                .padding(.bottom, AppTokens.Spacing.xxl)    // 对应 contentPadding = PaddingValues(bottom = ...)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity) // 对应 Modifier.fillMaxSize()
        .background(
            LinearGradient( // 对应 Brush.verticalGradient
                gradient: Gradient(colors: [
                    AppTokens.Colors.secondary.opacity(0.3), // 对应 .copy(alpha = 0.3f)
                    AppTokens.Colors.background,
                    AppTokens.Colors.primary.opacity(0.3)    // 对应 .copy(alpha = 0.3f)
                ]),
                startPoint: .top,
                endPoint: .bottom
            )
        )
        // Kotlin Column 的 Modifier.padding(AppTokens.Spacing.lg) 被分解为
        // 标题和 ScrollView 的水平填充，以及 VStack 的垂直 spacing，
        // 确保视觉效果一致。
    }
}

// MARK: - App Entry Point
// 翻译自 Kotlin 的 MainActivity 和 AppTheme，作为 SwiftUI 应用的入口。

@main
struct ItineraryPlannerApp: App {
    var body: some Scene {
        WindowGroup {
            ContentView()
        }
    }
}

struct ContentView: View {
    var body: some View {
        RootScreen()
            .ignoresSafeArea() // 对应 WindowCompat.setDecorFitsSystemWindows(window, false) 和 hide(WindowInsetsCompat.Type.systemBars())
            .statusBarHidden(true) // 对应 hide(WindowInsetsCompat.Type.systemBars())
    }
}

// MARK: - Preview
// 翻译自 Kotlin 的 PreviewRoot Composable，用于 Xcode 预览。

struct RootScreen_Previews: PreviewProvider {
    static var previews: some View {
        RootScreen()
    }
}

File: suc/kotlin/049.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Shapes
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.Typography
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "064*ItineraryPlanner*en"
private const val UI_TYPE = "Travel"
private const val STYLE_THEME = "Warm Gradient"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFFFF7043)
        val secondary = Color(0xFFFFB74D)
        val tertiary = Color(0xFFFFD180)
        val background = Color(0xFFFFF8F2)
        val surface = Color(0xFFFFFFFF)
        val surfaceVariant = Color(0xFFFFE0B2)
        val outline = Color(0xFFD7CCC8)
        val success = Color(0xFF43A047)
        val warning = Color(0xFFFFB300)
        val error = Color(0xFFE53935)
        val onPrimary = Color(0xFFFFFFFF)
        val onSecondary = Color(0xFF3E2723)
        val onTertiary = Color(0xFF3E2723)
        val onBackground = Color(0xFF3E2723)
        val onSurface = Color(0xFF3E2723)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(8.dp)
        val medium = RoundedCornerShape(14.dp)
        val large = RoundedCornerShape(20.dp)
    }
    object Spacing {
        val sm = 6.dp
        val md = 10.dp
        val lg = 16.dp
        val xl = 24.dp
        val xxl = 36.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.1f)
        val level2 = ShadowSpec(6.dp, 8.dp, 4.dp, 0.15f)
        val level3 = ShadowSpec(10.dp, 12.dp, 6.dp, 0.18f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

data class TripCard(val id: Int, val title: String, val days: Int, val price: String)

@OptIn(ExperimentalFoundationApi::class)
@Composable
fun RootScreen() {
    val trips = remember {
        listOf(
            TripCard(1, "Kyoto Cherry Trail", 4, "$460"),
            TripCard(2, "Tokyo City Break", 3, "$390"),
            TripCard(3, "Osaka Gourmet Tour", 5, "$520"),
            TripCard(4, "Mount Fuji Escape", 2, "$280"),
            TripCard(5, "Okinawa Beach Week", 6, "$740"),
            TripCard(6, "Hokkaido Winter Lights", 5, "$680")
        )
    }
    Scaffold(
        contentWindowInsets = androidx.compose.foundation.layout.WindowInsets(0),
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(
                    Brush.verticalGradient(
                        listOf(
                            AppTokens.Colors.secondary.copy(alpha = 0.3f),
                            AppTokens.Colors.background,
                            AppTokens.Colors.primary.copy(alpha = 0.3f)
                        )
                    )
                )
                .padding(AppTokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
        ) {
            Text("Itinerary Planner", style = AppTokens.TypographyTokens.display, color = AppTokens.Colors.primary)
            LazyVerticalGrid(
                columns = GridCells.Adaptive(160.dp),
                verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.lg),
                horizontalArrangement = Arrangement.spacedBy(AppTokens.Spacing.lg),
                contentPadding = PaddingValues(bottom = AppTokens.Spacing.xxl)
            ) {
                items(trips) { trip ->
                    Card(
                        shape = AppTokens.Shapes.large,
                        colors = CardDefaults.cardColors(containerColor = AppTokens.Colors.surface),
                        elevation = CardDefaults.cardElevation(AppTokens.ElevationMapping.level2.elevation)
                    ) {
                        Column(
                            modifier = Modifier
                                .fillMaxWidth()
                                .background(AppTokens.Colors.surface)
                                .padding(AppTokens.Spacing.md),
                            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm)
                        ) {
                            Box(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .height(100.dp)
                                    .background(AppTokens.Colors.surfaceVariant, AppTokens.Shapes.medium)
                            )
                            Text(trip.title, style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.onSurface)
                            Text("${trip.days} days", style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.onSurface.copy(alpha = 0.7f))
                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween,
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Text(trip.price, style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.primary)
                                Button(
                                    onClick = {},
                                    colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary, contentColor = AppTokens.Colors.onPrimary),
                                    shape = AppTokens.Shapes.medium,
                                    modifier = Modifier.height(36.dp)
                                ) { Text("Details", style = AppTokens.TypographyTokens.label) }
                            }
                        }
                    }
                }
            }
        }
    }
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFFF8F2)
@Composable
fun PreviewRoot() {
    AppTheme {
        RootScreen()
    }
}
